src/eval/context.rs

```diff
@@ -104,6 +104,17 @@ impl <'a, S: State + ?Sized> LoadedContext<'a, S> {
         r
     }
 
+    pub fn with_other_state<F, R>(&mut self, state: &mut &mut S, f: F) -> R
+    where F: FnOnce(&mut LoadedContext<'a, S>) -> R {
```

Comment:
Haha, it's a mutable reference to a mutable reference to a State!

Because [mem::swap](https://doc.rust-lang.org/stable/std/mem/fn.swap.html) doesn't allow for unsized types (and our State is allowed to be unsized), we can't swap with just an `&mut S`.  Instead, we need to literally swap the pointer to the pointer.

One example of this would be if we had `LoadedContext<MyTrait>`.  This is legal right now because we allow contexts to be loaded with references to unsized types.  However, you can't do `mem::swap` with `&mut MyTrait, &mut MyTrait` because the underlying representations could be different.  After all, the only thing we know is that there is a mutable reference to *something* that implements `MyTrait`.

I agree that it's a bit weird.